stdlib/public/Concurrency/TaskSleep.swift

```diff
@@ -105,25 +105,55 @@ extension Task where Success == Never, Failure == Never {
     }
   }
 
+  /// A simple wrapper for a pointer to heap allocated storage of a `SleepState`
+  /// value. This wrapper is `Sendable` because it facilitates atomic load and
+  /// exchange operations on the underlying storage. However, this wrapper is also
+  /// _unsafe_ because the owner must manually deallocate the token once it is no
+  /// longer needed.
+  struct UnsafeSleepStateToken: @unchecked Sendable {
+    let wordPtr: UnsafeMutablePointer<Builtin.Word>
+
+    /// Allocates the underlying storage and sets the value to `.notStarted`.
+    init() {
+      wordPtr = .allocate(capacity: 1)
+      Builtin.atomicstore_seqcst_Word(
+          wordPtr._rawValue, SleepState.notStarted.word._builtinWordValue)
+    }
+
+    /// Atomically loads the current state.
+    func load() -> SleepState {
+      return SleepState(word: Builtin.atomicload_seqcst_Word(wordPtr._rawValue))
+    }
+
+    /// Attempts to atomically set the stored value to `desired` if the current
+    /// value is equal to `expected`. Returns true if the exchange was successful.
+    func exchange(expected: SleepState, desired: SleepState) -> Bool {
+      let (_, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+          wordPtr._rawValue,
+          expected.word._builtinWordValue,
+          desired.word._builtinWordValue)
+      return Bool(_builtinBooleanLiteral: won)
+    }
+
+    /// Deallocates the underlying storage.
+    func deallocate() {
+      wordPtr.deallocate()
+    }
+  }
+
   /// Called when the sleep(nanoseconds:) operation woke up without being
   /// canceled.
-  static func onSleepWake(
-      _ wordPtr: UnsafeMutablePointer<Builtin.Word>
-  ) {
+  static func onSleepWake(_ token: UnsafeSleepStateToken) {
     while true {
-      let state = SleepState(loading: wordPtr)
+      let state = token.load()
```

Comment:
late to the party here -- I was looking into this recently and it turned out we didn't properly handle `nonisolated(unsafe)` for local variables (heh). That was fixed since then so I wondered if we could just `nonisolated(unsafe)` the `wordPtr`  now and avoid the wrapper entirely.

Although avoiding the alloc entirely would be even better...! So thanks for investigating that too.